serif_swing/src/main/kotlin/serif_swing/App.kt

```diff
@@ -119,63 +134,106 @@ class SwingChatRoom(val transition: (MatrixState, Boolean) -> Unit, val panel: J
-        redrawMessages()
-        panel.add(JScrollPane(inner_scroll_pane,
-                              JScrollPane.VERTICAL_SCROLLBAR_ALWAYS,
-                              JScrollPane.HORIZONTAL_SCROLLBAR_NEVER),
-                  BorderLayout.CENTER)
+        redrawMessages(last_window_width)
+        panel.add(
+            JScrollPane(
+                inner_scroll_pane,
+                JScrollPane.VERTICAL_SCROLLBAR_ALWAYS,
+                JScrollPane.HORIZONTAL_SCROLLBAR_NEVER
+            ),
+            BorderLayout.CENTER
+        )
 
         val message_panel = JPanel()
         message_panel.layout = BorderLayout()
         var back_button = JButton("Back")
         message_panel.add(back_button, BorderLayout.LINE_START)
         message_panel.add(message_field, BorderLayout.CENTER)
+        val msg_panel_actions = JPanel()
+        msg_panel_actions.layout = BoxLayout(msg_panel_actions, BoxLayout.LINE_AXIS)
+        var attach_button = JButton("+")
+        msg_panel_actions.add(attach_button)
         var send_button = JButton("Send")
-        message_panel.add(send_button, BorderLayout.LINE_END)
+        msg_panel_actions.add(send_button)
+        message_panel.add(msg_panel_actions, BorderLayout.LINE_END)
         panel.add(message_panel, BorderLayout.PAGE_END)
         val onSend: (ActionEvent) -> Unit = {
             val text = message_field.text
             message_field.text = ""
             transition(m.sendMessage(text), true)
         }
+        val onAttach: (ActionEvent) -> Unit = {
+            val fc = JFileChooser()
+            val iff = ImageFileFilter()
+            fc.addChoosableFileFilter(iff)
+            fc.setFileFilter(iff)
+            val ret = fc.showDialog(panel, "Attach")
+            if(ret == JFileChooser.APPROVE_OPTION) {
+                val file = fc.getSelectedFile()
+                message_field.text = ""
+                transition(m.sendMessage(file.toPath().toString(), is_img = true), true)
+                println("Selected ${file.toPath()}")
+            }
+        }
         message_field.addActionListener(onSend)
         send_button.addActionListener(onSend)
+        attach_button.addActionListener(onAttach)
         back_button.addActionListener({ transition(m.exitRoom(), true) })
-
     }
-    fun redrawMessages() {
+    fun redrawMessages(draw_width: Int) {
         inner_scroll_pane.removeAll()
         val layout = inner_scroll_pane.layout as GroupLayout
         val parallel_group = layout.createParallelGroup(GroupLayout.Alignment.LEADING)
         var seq_vert_groups = layout.createSequentialGroup()
-        for ((sender, message) in m.messages) {
-            val sender = JLabel("$sender:  ")
-
-            val message = JTextArea(message)
-            message.setEditable(false)
-            message.lineWrap = true
-            message.wrapStyleWord = true
+        for (msg in m.messages) {
+            val _sender = msg.sender
```

Comment:
Hmm, this shouldn't have an underscore, when did it get here?

Reply:
I added this to get rid of the shadowing warning regarding the text area down below.